File: web/src/model/victron_controller/dashboard/SocProjectionKind.ts
// @ts-nocheck
import {BaboonCodecContext, BaboonBinWriter, BinTools, BaboonBinReader, Lazy} from '../../BaboonSharedRuntime'

export enum SocProjectionKind {
    Natural = "Natural",
    Idle = "Idle",
    ScheduledCharge = "ScheduledCharge",
    FullChargePush = "FullChargePush",
    Clamped = "Clamped",
    SolarCharge = "SolarCharge",
    Drain = "Drain",
    ForcedNoExport = "ForcedNoExport",
    PreserveForZappi = "PreserveForZappi",
    BelowExportThreshold = "BelowExportThreshold",
    EveningDischarge = "EveningDischarge",
    BatteryFull = "BatteryFull"
}

export const SocProjectionKind_values: ReadonlyArray<SocProjectionKind> = [
    SocProjectionKind.Natural,
    SocProjectionKind.Idle,
    SocProjectionKind.ScheduledCharge,
    SocProjectionKind.FullChargePush,
    SocProjectionKind.Clamped,
    SocProjectionKind.SolarCharge,
    SocProjectionKind.Drain,
    SocProjectionKind.ForcedNoExport,
    SocProjectionKind.PreserveForZappi,
    SocProjectionKind.BelowExportThreshold,
    SocProjectionKind.EveningDischarge,
    SocProjectionKind.BatteryFull
] as const;

export function SocProjectionKind_parse(s: string): SocProjectionKind {
    const found = SocProjectionKind_values.find(v => v === s);
    if (found === undefined) {
        throw new Error("Unknown SocProjectionKind variant: " + s);
    }
    return found;
}

export class SocProjectionKind_UEBACodec {
    public encode(ctx: BaboonCodecContext, value: SocProjectionKind, writer: BaboonBinWriter): unknown {
        if (this !== SocProjectionKind_UEBACodec.lazyInstance.value) {
          return SocProjectionKind_UEBACodec.lazyInstance.value.encode(ctx, value, writer)
        }
    
        switch (value) {
            case "Natural": BinTools.writeByte(writer, 0); break;
                case "Idle": BinTools.writeByte(writer, 1); break;
                case "ScheduledCharge": BinTools.writeByte(writer, 2); break;
                case "FullChargePush": BinTools.writeByte(writer, 3); break;
                case "Clamped": BinTools.writeByte(writer, 4); break;
                case "SolarCharge": BinTools.writeByte(writer, 5); break;
                case "Drain": BinTools.writeByte(writer, 6); break;
                case "ForcedNoExport": BinTools.writeByte(writer, 7); break;
                case "PreserveForZappi": BinTools.writeByte(writer, 8); break;
                case "BelowExportThreshold": BinTools.writeByte(writer, 9); break;
                case "EveningDischarge": BinTools.writeByte(writer, 10); break;
                case "BatteryFull": BinTools.writeByte(writer, 11); break;
            default: throw new Error("Unknown enum variant: " + value);
        }
    }
    
    public decode(ctx: BaboonCodecContext, reader: BaboonBinReader): SocProjectionKind {
        if (this !== SocProjectionKind_UEBACodec .lazyInstance.value) {
            return SocProjectionKind_UEBACodec.lazyInstance.value.decode(ctx, reader)
        }
    
        const tag = BinTools.readByte(reader);
        switch (tag) {
            case 0: return "Natural" as SocProjectionKind;
                case 1: return "Idle" as SocProjectionKind;
                case 2: return "ScheduledCharge" as SocProjectionKind;
                case 3: return "FullChargePush" as SocProjectionKind;
                case 4: return "Clamped" as SocProjectionKind;
                case 5: return "SolarCharge" as SocProjectionKind;
                case 6: return "Drain" as SocProjectionKind;
                case 7: return "ForcedNoExport" as SocProjectionKind;
                case 8: return "PreserveForZappi" as SocProjectionKind;
                case 9: return "BelowExportThreshold" as SocProjectionKind;
                case 10: return "EveningDischarge" as SocProjectionKind;
                case 11: return "BatteryFull" as SocProjectionKind;
            default: throw new Error("Unknown enum variant tag: " + tag);
        }
    }

    public static readonly BaboonDomainVersion = '0.3.0'
    public baboonDomainVersion() {
        return SocProjectionKind_UEBACodec.BaboonDomainVersion
    }
    public static readonly BaboonDomainIdentifier = 'victron_controller.dashboard'
    public baboonDomainIdentifier() {
        return SocProjectionKind_UEBACodec.BaboonDomainIdentifier
    }
    public static readonly BaboonTypeIdentifier = 'victron_controller.dashboard/:#SocProjectionKind'
    public baboonTypeIdentifier() {
        return SocProjectionKind_UEBACodec.BaboonTypeIdentifier
    }

    protected static lazyInstance = new Lazy(() => new SocProjectionKind_UEBACodec())
    public static get instance(): SocProjectionKind_UEBACodec {
        return SocProjectionKind_UEBACodec.lazyInstance.value
    }
}

File: crates/dashboard-model/src/victron_controller/dashboard/v0_2_0/soc_projection_kind.rs


#[derive(Clone, Debug, PartialEq, Eq, Hash, PartialOrd, Ord, serde::Serialize, serde::Deserialize)]
pub enum SocProjectionKind {
    Natural,
    Idle,
    ScheduledCharge,
    FullChargePush,
    Clamped,
    SolarCharge,
    Drain,
    ForcedNoExport,
    PreserveForZappi,
    BelowExportThreshold,
    EveningDischarge,
    BatteryFull,
}

impl SocProjectionKind {
    pub fn parse(s: &str) -> Result<Self, String> {
        match s {
            "Natural" => Ok(SocProjectionKind::Natural),
            "Idle" => Ok(SocProjectionKind::Idle),
            "ScheduledCharge" => Ok(SocProjectionKind::ScheduledCharge),
            "FullChargePush" => Ok(SocProjectionKind::FullChargePush),
            "Clamped" => Ok(SocProjectionKind::Clamped),
            "SolarCharge" => Ok(SocProjectionKind::SolarCharge),
            "Drain" => Ok(SocProjectionKind::Drain),
            "ForcedNoExport" => Ok(SocProjectionKind::ForcedNoExport),
            "PreserveForZappi" => Ok(SocProjectionKind::PreserveForZappi),
            "BelowExportThreshold" => Ok(SocProjectionKind::BelowExportThreshold),
            "EveningDischarge" => Ok(SocProjectionKind::EveningDischarge),
            "BatteryFull" => Ok(SocProjectionKind::BatteryFull),
            _ => Err(format!("Unknown variant: {}", s)),
        }
    }

    pub fn all() -> Vec<SocProjectionKind> {
        vec![
            SocProjectionKind::Natural,
            SocProjectionKind::Idle,
            SocProjectionKind::ScheduledCharge,
            SocProjectionKind::FullChargePush,
            SocProjectionKind::Clamped,
            SocProjectionKind::SolarCharge,
            SocProjectionKind::Drain,
            SocProjectionKind::ForcedNoExport,
            SocProjectionKind::PreserveForZappi,
            SocProjectionKind::BelowExportThreshold,
            SocProjectionKind::EveningDischarge,
            SocProjectionKind::BatteryFull,
        ]
    }
}

impl std::fmt::Display for SocProjectionKind {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            SocProjectionKind::Natural => write!(f, "Natural"),
            SocProjectionKind::Idle => write!(f, "Idle"),
            SocProjectionKind::ScheduledCharge => write!(f, "ScheduledCharge"),
            SocProjectionKind::FullChargePush => write!(f, "FullChargePush"),
            SocProjectionKind::Clamped => write!(f, "Clamped"),
            SocProjectionKind::SolarCharge => write!(f, "SolarCharge"),
            SocProjectionKind::Drain => write!(f, "Drain"),
            SocProjectionKind::ForcedNoExport => write!(f, "ForcedNoExport"),
            SocProjectionKind::PreserveForZappi => write!(f, "PreserveForZappi"),
            SocProjectionKind::BelowExportThreshold => write!(f, "BelowExportThreshold"),
            SocProjectionKind::EveningDischarge => write!(f, "EveningDischarge"),
            SocProjectionKind::BatteryFull => write!(f, "BatteryFull"),
        }
    }
}

impl std::str::FromStr for SocProjectionKind {
    type Err = String;
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        Self::parse(s)
    }
}

impl crate::baboon_runtime::BaboonBinCodecIndexed for SocProjectionKind {
    fn index_elements_count(_ctx: &crate::baboon_runtime::BaboonCodecContext) -> u16 {
        0
    }
}

impl crate::baboon_runtime::BaboonBinEncode for SocProjectionKind {
    fn encode_ueba(&self, _ctx: &crate::baboon_runtime::BaboonCodecContext, writer: &mut dyn std::io::Write) -> std::io::Result<()> {
        match self {
            SocProjectionKind::Natural => crate::baboon_runtime::bin_tools::write_byte(writer, 0)?,
            SocProjectionKind::Idle => crate::baboon_runtime::bin_tools::write_byte(writer, 1)?,
            SocProjectionKind::ScheduledCharge => crate::baboon_runtime::bin_tools::write_byte(writer, 2)?,
            SocProjectionKind::FullChargePush => crate::baboon_runtime::bin_tools::write_byte(writer, 3)?,
            SocProjectionKind::Clamped => crate::baboon_runtime::bin_tools::write_byte(writer, 4)?,
            SocProjectionKind::SolarCharge => crate::baboon_runtime::bin_tools::write_byte(writer, 5)?,
            SocProjectionKind::Drain => crate::baboon_runtime::bin_tools::write_byte(writer, 6)?,
            SocProjectionKind::ForcedNoExport => crate::baboon_runtime::bin_tools::write_byte(writer, 7)?,
            SocProjectionKind::PreserveForZappi => crate::baboon_runtime::bin_tools::write_byte(writer, 8)?,
            SocProjectionKind::BelowExportThreshold => crate::baboon_runtime::bin_tools::write_byte(writer, 9)?,
            SocProjectionKind::EveningDischarge => crate::baboon_runtime::bin_tools::write_byte(writer, 10)?,
            SocProjectionKind::BatteryFull => crate::baboon_runtime::bin_tools::write_byte(writer, 11)?,
        }
        Ok(())
    }
}

impl crate::baboon_runtime::BaboonBinDecode for SocProjectionKind {
    fn decode_ueba(_ctx: &crate::baboon_runtime::BaboonCodecContext, reader: &mut dyn std::io::Read) -> Result<Self, Box<dyn std::error::Error>> {
        let tag = crate::baboon_runtime::bin_tools::read_byte(reader)?;
        match tag {
            0 => Ok(SocProjectionKind::Natural),
            1 => Ok(SocProjectionKind::Idle),
            2 => Ok(SocProjectionKind::ScheduledCharge),
            3 => Ok(SocProjectionKind::FullChargePush),
            4 => Ok(SocProjectionKind::Clamped),
            5 => Ok(SocProjectionKind::SolarCharge),
            6 => Ok(SocProjectionKind::Drain),
            7 => Ok(SocProjectionKind::ForcedNoExport),
            8 => Ok(SocProjectionKind::PreserveForZappi),
            9 => Ok(SocProjectionKind::BelowExportThreshold),
            10 => Ok(SocProjectionKind::EveningDischarge),
            11 => Ok(SocProjectionKind::BatteryFull),
            _ => Err(format!("Unknown enum variant tag: {}", tag).into()),
        }
    }
}